src/services/AuthProvider.js

```diff
@@ -0,0 +1,72 @@
+import React, { useContext, useState, useEffect } from 'react';
+import { auth, db, googleProvider } from './FirebaseService';
+
+const AuthContext = React.createContext();
+
+export function useAuth() {
+  return useContext(AuthContext);
+}
+
+// eslint-disable-next-line react/prop-types
+export function AuthProvider({ children }) {
+  const [currentUser, setCurrentUser] = useState();
+  const [loading, setLoading] = useState(true);
+
+  async function emailLogin(email, password) {
+    return auth.signInWithEmailAndPassword(email, password);
+  }
+
+  async function signup(email, password, uname) {
+    return auth.createUserWithEmailAndPassword(email, password).then((userData) => {
+      emailLogin(email, password).then(() => {
+        db.collection('users').doc(userData.user.uid).set({
+          username: uname,
+          bookmarks: [],
+          folderNames: [],
+        });
+      });
+    });
+  }
+
+  async function googleLogin() {
+    return auth.signInWithPopup(googleProvider);
+  }
+
+  async function logout() {
+    return auth.signOut();
+  }
+
+  async function resetPassword(email) {
+    return auth.sendPasswordResetEmail(email);
+  }
+
+  async function updateEmail(email) {
+    return currentUser.updateEmail(email);
+  }
+
+  async function updatePassword(password) {
+    return currentUser.updatePassword(password);
+  }
+
+  useEffect(() => {
+    const unsubscribe = auth.onAuthStateChanged((user) => {
+      setCurrentUser(user);
+      setLoading(false);
+    });
+
+    return unsubscribe;
```

Comment:
Note whenever you place something as a return on a `useEffect` hook, you also call that when the component unmounts. Usually return values in `useEffect`s are only for cleanups and teardowns of code when a component unmounts. Refer to this link and look at the `clean up` section, this will describe what the return does in `useEffect` hooks [React docs](https://reactjs.org/docs/hooks-effect.html)

Reply:
Hi kuya tony, I'm having doubts on removing this unsubscribe and I currently have no idea on implementing it in other way as well. My intention in putting that code is to unsubscribe to the listener from firebase auth's onAuthStateChanged when the app is not active anymore. Can I get further explanation on this po? Thank youu